src/charm.py

```diff
@@ -84,18 +75,18 @@ def generate_worker_layer(self, worker: Worker) -> Layer:
 
         This method assumes that worker.roles is valid.
         """
+
         roles = worker.roles
         if not len(roles) == 1:
             raise RolesConfigurationError(
                 f"Worker can only have 1 role configured. {len(roles)} found."
             )
         role = roles[0]
-        if role not in self._valid_roles:
-            raise RolesConfigurationError(
-                f"Invalid role {role}. Should be one of {self._valid_roles}"
-            )
         if role == "all":
             role = "scalable-single-binary"
+
+        if role == "scalable-single-binary" or role == "metrics-generator":
```

Comment:
this side-effecting feels bad practice to me. This method is supposed to generate a Layer, why is it patching config files?

More generally, writing the config is the Worker object's responsibility, and here we're crossing that boundary.
What we actually would like is for the Worker to give us a chance to modify the configuration instead of pushing it as-is to the container. Also, changing the config twice instead of once might be bad as it could cause restarts, downtime, and who knows what.

So, a better solution to this situation is, I think, to do this in the Worker object instead. 

--> https://github.com/canonical/cos-lib/pull/65